src/Lean/Message.lean

```diff
@@ -94,7 +108,10 @@ def isEmpty : MessageData → Bool
   | _ => false
 
 variable (p : Name → Bool) in
-/-- Returns true when the message contains a `MessageData.tagged tag ..` constructor where `p tag` is true. -/
+/-- Returns true when the message contains a `MessageData.tagged tag ..` constructor where `p tag`
+is true.
+This does not descend into lazily generated subtress (`.ofLazy`).
```

Comment:
Why is `false` the right approximation here? I would think that lazy messages absolutely could contain tags, and I'd expect that this was used to make some operation go faster by ignoring them.

Reply:
Also a good question. Note that this function doesn't check for just any tag, but for specific ones. Typical uses are

```
      -- discard elaboration errors, except for a few important and unlikely misleading ones, on
      -- parse error
      msgs := ⟨msgs.msgs.filter fun msg =>
        msg.data.hasTag (fun tag => tag == `Elab.synthPlaceholder ||
          tag == `Tactic.unsolvedGoals || (`_traceMsg).isSuffixOf tag)⟩
```
…
```
def _root_.Lean.MessageData.isDeprecationWarning (msg : MessageData) : Bool :=
  msg.hasTag (· == ``deprecatedAttr)
```
…
```
/-- Returns true if this is a message produced by the unused variable linter.
This is used to give these messages an additional "faded" style in the editor. -/
def MessageData.isUnusedVariableWarning (msg : MessageData) : Bool :=
  msg.hasTag (· == Linter.linter.unusedVariables.name)
```

I assume that these tags are always somewhere near the root of the message, and not inside a `.lazy` constructor, e.g. when created like this:
```
def logLint [Monad m] [MonadLog m] [AddMessageContext m] [MonadOptions m]
    (linterOption : Lean.Option Bool) (stx : Syntax) (msg : MessageData) : m Unit :=
  let disable := m!"note: this linter can be disabled with `set_option {linterOption.name} false`"
  logWarningAt stx (.tagged linterOption.name m!"{msg}\n{disable}")
```

One could imagine more verbose types to track such invariants, but I’d shy away from such a refactoring (I believe the data type already has such invariants, e.g. that certain constructors are always inside an `.withContext`).

Maybe good enough just documenting this assumption here?

```suggestion

This does not descend into lazily generated subtress (`.ofLazy`); message tags
of interest(like those added by `logLinter`) are expected to be near the root
of the `MessageData`, and not hidden inside `.ofLazy`.
```